File: python-engine/routes/instruments.py
from fastapi import APIRouter, Query
from typing import List
from models import Instrument, ExpiryInfo, ExpiriesResponse

router = APIRouter()


@router.get("/search")
async def search_instruments(q: str = Query("", description="Search query")):
    """Search instruments by human-readable query like 'NIFTY 23900 CE'

    Uses the Upstox SDK InstrumentsApi.search_instrument() for live search.
    Returns results in the format expected by the frontend search component.
    """
    from main import engine
    if not q or len(q) < 2:
        return {"results": []}

    results = await engine.search_instruments_async(q)
    return {"results": results}


@router.get("/expiries", response_model=ExpiriesResponse)
async def get_expiries(underlying: str = Query("NIFTY")):
    """Get available expiry dates for an underlying"""
    from main import engine
    expiries = await engine.get_expiries_async(underlying)
    return {"underlying": underlying, "expiries": expiries}

File: mini-services/trading-engine/routes/instruments.ts
/**
 * Instrument search and expiry routes.
 */

import { Router, type Request, type Response } from 'express';
import { getSimulator } from '../market-simulator';

const router = Router();

/**
 * GET /api/instruments/search?q={query}
 * Search instruments by human-readable query (e.g., 'NIFTY 23900 CE 25 Jun 2026')
 */
router.get('/search', (req: Request, res: Response) => {
  const q = (req.query.q as string) || '';
  const simulator = getSimulator();
  const results = simulator.searchInstruments(q);
  res.json({ instruments: results, total: results.length });
});

/**
 * GET /api/instruments/expiries?underlying={symbol}
 * Get current week expiry dates for an underlying.
 */
router.get('/expiries', (req: Request, res: Response) => {
  const underlying = (req.query.underlying as string) || 'NIFTY';
  const simulator = getSimulator();
  const expiries = simulator.getExpiries(underlying);
  res.json({ underlying, expiries });
});

export default router;
